staging/src/k8s.io/apiserver/pkg/server/options/authentication.go

```diff
@@ -116,6 +120,9 @@ func (s *RequestHeaderAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringSliceVar(&s.UsernameHeaders, "requestheader-username-headers", s.UsernameHeaders, ""+
 		"List of request headers to inspect for usernames. X-Remote-User is common.")
 
+	fs.StringSliceVar(&s.UIDHeaders, "requestheader-uid-headers", s.UIDHeaders, ""+
+		"List of request headers to inspect for UIDs. X-Remote-Uid is suggested.")
```

Comment:
Per triage meeting review from today:

This flag needs to not be wired up in the first release that contains this feature as then it would be possible for old KAS to proxy a request to a new aggregated API server without dropping the UID header.  Everything else is safe to do in this PR.  To be able to test this code path in integration tests, we can use an approach similar to `SetServiceResolverForTests` to let us set the `UIDHeaders` field.  A future PR in a later release would simply drop that global in place of the actual flag wiring.

I believe it is safe to update `NewDelegatingAuthenticationOptions` with the `X-Remote-Uid` as a default because no matter what that relies on the front proxy dropping headers that the client should not set.  And that code path isn't really used because the vast majority of cases should be using the config map, which will only get updated once KAS is updated and the `--requestheader-uid-headers` CLI flag is set.  @liggitt any thoughts on the upgrade story here?

Once we let people configure this flag on KAS, we should require that it contains at least `X-Remote-Uid` so that API aggregation will work correctly (we can be stricter here because the config is new).

Reply:
Just got here, need to think about the skew implications.

There was just a lengthy discussion about what skew between kube-apiserver and aggregated servers are allowed / recommended / supported (https://github.com/kubernetes/website/pull/46109, https://github.com/kubernetes/kubernetes/issues/124533, https://github.com/kubernetes/kubernetes/issues/124655, api-machinery meeting https://www.youtube.com/watch?v=0TXm-DGcK1k&t=31m). We don't really dictate whether aggregated API servers' use of k8s.io/apiserver library can / should be newer than kube-apiserver today. cc @deads2k as well who might have thoughts.

I agree our defaults (in kube-apiserver and k8s.io/apiserver library) should not lead to aggregated API servers trusting headers _before_ kube-apiserver sets / guards / protects them.

I think the following defaults work:
* kube-apiserver 1.31+ does not default the uid CLI flag (just like the other requestheader flags)
* kube-apiserver 1.31+ (as kube-aggregator) unconditionally clears the standard uid header when proxying (just like it clears the other standard headers)
* kube-apiserver 1.31+ (as kube-aggregator) unconditionally sets the standard uid header when proxying (just like it sets the other standard headers)
* k8s.io/apiserver library v0.31.0+ wires the uid CLI flag
* k8s.io/apiserver library v0.31.0+ NewDelegatingAuthenticationOptions() does not default the uid header; this keeps the CLI flags safe against much older kube-apiservers which aren't protecting the header; in practice, most aggregated API servers pull their auth header config from the configmap published by kube-apiserver rather than the CLI flags, so if/when their kube-apiserver publishes uid header config, they'll start using it

In the future (kube-apiserver 1.32+) we can consider making kube-apiserver also include the standard username/groups/extra/uid headers the aggregator uses (which are not configurable) in the configmap it publishes containing auth config, so that aggregation works even if kube-apiserver only set non-standard requestheader flag options

Reply:
I removed the defaulting from the `DelegatingAuthenticationOptions()` which I think was the only thing missing to get us to the desired state as described by Jordan.

We'll probably still have to think how to update the cluster trust CM - https://github.com/kubernetes/kubernetes/pull/115834#discussion_r1611467973.